### PR TITLE
relogin

### PR DESCRIPTION
**Pull request by chrisnojima**
This PR got big so i'm publishing it here now, i'll likely do some more testing or relogin.

console.error is now console.warn, else we'll get redboxes everywhere
had to fix how we init the daemon so we get notifications working, we really should revisit how login nav / bootstraping / etc works w/ sagas later

### CODE REVIEW

File: desktop/app/index.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ if (__DEV__) { // eslint-disable-line no-undef
   app.commandLine.appendSwitch('v', 3)
 }
 
-hello(process.pid, 'Main Thread', process.argv)
+hello(process.pid, 'Main Thread', process.argv, __VERSION__) // eslint-disable-line no-undef
```
Comment (chrisnojima) [inline desktop/app/index.js]: i pulled __VERSION__ out of hello and made it a param so desktop could pass it in. i'll ticket us plumbing this through on mobile

File: go/loopback/main.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ import (
 var con net.Conn
 var startOnce sync.Once
 
-// ServerURI should match run mode environment.
+// Init ServerURI should match run mode environment.
```
Comment (chrisnojima) [inline go/loopback/main.go]: go lint complains

File: go/loopback/main.go [hunk 3]
```diff
@@ -35,12 +35,14 @@ func Init(homeDir string, runModeStr string, serverURI string, accessGroupOverri
 		if err != nil {
 			panic(err)
 		}
-		(service.NewService(g, false)).StartLoopbackServer()
+		service := (service.NewService(g, false))
+		service.StartLoopbackServer()
+		service.G().SetService()
```
Comment (chrisnojima) [inline go/loopback/main.go]: we need this else we don't init part of the daemon

File: shared/native/notification-listeners.native.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import {bootstrap} from '../actions/config'
+import {logoutDone} from '../actions/login'
+import type {Dispatch} from '../constants/types/flux'
+import type {incomingCallMapType} from '../constants/types/flow-types'
+
+// Keep track of the last time we notified and ignore if its the same
+let lastLoggedInNotifyUsername = null
+
+// TODO(mm) Move these to their own actions
+export default function (dispatch: Dispatch, getState: () => Object, notify: any): incomingCallMapType {
```
Comment (chrisnojima) [inline shared/native/notification-listeners.native.js]: we just handle login/logout for now

File: shared/nav.ios.js [hunk 5]
```diff
@@ -81,6 +84,15 @@ class Nav extends Component {
     super(props)
 
     this.props.bootstrap()
+    this.props.listenForNotifications()
+
+    // Handle logUi.log
+    ListenLogUi()
+
+    // Introduce ourselves to the service
+    hello(0, 'iOS app', [], '0.0.0') // TODO real version
```
Comment (chrisnojima) [inline shared/nav.ios.js]: i made a ticket for this already

File: shared/nav.ios.js [hunk 6]
```diff
@@ -81,6 +84,15 @@ class Nav extends Component {
...
+
+    // TODO android also!
```
Comment (chrisnojima) [inline shared/nav.ios.js]: i'll work on this now also

File: shared/reducers/profile.js [hunk 7]
```diff
@@ -9,7 +9,7 @@ export default function (state = initialState, action) {
 
   switch (action.type) {
     case CommonConstants.resetStore:
-      return {...initialState}
+      return Immutable.fromJS({...initialState.toJS()})
```
Comment (chrisnojima) [inline shared/reducers/profile.js]: so react native doens't do object assign perfectly so it balks at this kind of stuff

**Comment by chrisnojima** [comment]
@keybase/react-hackers

File: shared/native/notification-listeners.native.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        getState().config.status &&
+        getState().config.status.loggedIn) {
+        notify('Logged out of Keybase')
+        dispatch(logoutDone())
+      }
+    },
+    'keybase.1.NotifySession.loggedIn': ({username}, response) => {
+      if (lastLoggedInNotifyUsername !== username) {
+        lastLoggedInNotifyUsername = username
+        notify('Logged in to Keybase as: ' + username)
```
Comment (MarcoPolo) [inline shared/native/notification-listeners.native.js]: Maybe have `native/notifications.native.js` `console.log` for now just so we can smoke test?

File: shared/reducers/profile.js (hunk 7 above)
Comment (MarcoPolo) [inline shared/reducers/profile.js]: I'm sorry, That's annoying.

How about: `initialState.asMutable().asImmutable()`

Matches the property we're looking for, and will probably not subtly loose data.
```
a.asMutable().asImmutable() === a
// => false
a.asMutable().asImmutable().equals(a)
// => true
```

File: shared/reducers/search.js [hunk 9]
```diff
@@ -20,7 +20,7 @@ const initialState: SearchState = Immutable.Map()
 
 export default function (state: SearchState = initialState, action: any): SearchState {
   if (action.type === CommonConstants.resetStore) {
-    return {...initialState}
+    return Immutable.fromJS({...initialState.toJS()})
```
Comment (MarcoPolo) [inline shared/reducers/search.js]: same as profile.

**Comment by MarcoPolo** [comment]
Looks good % C

are you adding the android `hello`? feel free to do that in another PR if it helps

**Comment by chrisnojima** [comment]
waiting on android until ios is totally finished: https://keybase.atlassian.net/browse/DESKTOP-954

**Pull request merged**